src/rgw/rgw_cacherequest.h

```diff
@@ -0,0 +1,106 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#ifndef RGW_CACHEREQUEST_H
+#define RGW_CACHEREQUEST_H
+
+#include <aio.h>
+
+#include "include/rados/librados.hpp"
+#include "include/Context.h"
+
+#include "common/ceph_mutex.h"
+
+class CacheRequest {
+  public:
+    ceph::mutex lock = ceph::make_mutex("CacheRequest");
+    int sequence;
+    bufferlist *pbl;
+    struct get_obj_data *op_data;
+    std::string oid;
+    off_t ofs;
+    off_t len;
+    librados::AioCompletion *lc;
+    std::string key;
+    off_t read_ofs;
+    Context *onack;
+    CephContext *cct;
+    CacheRequest(CephContext *_cct) : sequence(0), pbl(NULL), op_data(NULL), ofs(0), lc(NULL), read_ofs(0), cct(_cct) {};
+    virtual ~CacheRequest(){};
+    virtual void release()=0;
+    virtual void cancel_io()=0;
+    virtual int status()=0;
+    virtual void finish()=0;
+};
+
+struct L1CacheRequest : public CacheRequest{
+  int stat;
+  struct aiocb *paiocb;
+  L1CacheRequest(CephContext *_cct) :  CacheRequest(_cct), stat(-1), paiocb(NULL) {}
+  ~L1CacheRequest(){}
+  void release (){
+    lock.lock();
+    free((void *)paiocb->aio_buf);
```

Comment:
I'm surprised to see a free() here;  is a buffer being allocated w/malloc?  let's use new/delete at worst

Reply:
aio_buf is volatile void *. Would it still be okay to use delete()?

Reply:
1) C allocators aren't safer than C++; 2) why is it volatile, anyway; 3) in C and C++, memory can only be release when its safe to do so, I presume that's so here

Matt

Reply:
the memory allocation/deallocations using malloc and free here are following the documentation examples in `man 7 aio` and the volatile originates from `/usr/include/aio.h`
```
struct aiocb
{
  ...
  volatile void *aio_buf;	/* Location of buffer.  */
  ...
```